plugins/processors/enum/enum.go

```diff
@@ -0,0 +1,110 @@
+package enum
+
+import (
+	"strconv"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+var sampleConfig = `
+## NOTE This processor will map metric values to different values. It is aimed
+## to map enum values to numeric values.
+
+## Fields to be considered
+# [[processors.enum.fields]]
+#
+# Name of the field
+#   key = "name"
+#
+# Value Mapping Table
+#   [processors.enum.value_mappings]
+#     value1 = 1
+#     value2 = 2
+#
+## Alternatively the mapping table can be given in inline notation
+#   value_mappings = {value1 = 1, value2 = 2}
+`
+
+type EnumMapper struct {
+	Fields []Mapping
+}
+
+type Mapping struct {
+	Key           string
+	ValueMappings map[string]interface{}
+}
+
+func (mapper *EnumMapper) SampleConfig() string {
+	return sampleConfig
+}
+
+func (mapper *EnumMapper) Description() string {
+	return "Map enum values according to given table."
+}
+
+func (mapper *EnumMapper) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	out := make([]telegraf.Metric, 0, len(in))
+	for _, source := range in {
+		target, error := mapper.applyMappings(source)
+		if error == nil {
+			out = append(out, target)
+		} else {
+			out = append(out, source)
+		}
+	}
+	return out
+}
+
+func (mapper *EnumMapper) applyMappings(source telegraf.Metric) (telegraf.Metric, error) {
+	if fields, changed := mapper.applyFieldMappings(source.Fields()); changed == true {
```

Comment:
Instead of grabbing all the fields, it might be faster to use `HasField` to check for the fields we are interested in and modify the metric instead of creating a new one.

Reply:
Checking with `HasField`first is a great suggestion I will implement that. As discussed above I have trouble when modifying the metric directly in the case where it has just the one metric we are interested in. Note, that the implementation will only create a new metric when a substitution was made and pass on the original object otherwise.

Reply:
I flipped through the other comments and couldn't find the discussion of this, can you tell me more about the problem modifying the metric directly or point me in the right direction?

Reply:
I can provide a simple test to demonstrate my observations:

```go
import (
	"fmt"
	"testing"
	"time"

	"github.com/influxdata/telegraf/metric"
)

func TestDirectModification(t *testing.T) {
	metric, _ := metric.New("m",
		map[string]string{},
		map[string]interface{}{"test": "value"},
		time.Now(),
	)
	fmt.Println(metric.String()) // original state
	metric.RemoveField("test")
	fmt.Println(metric.String()) // field is not removed
	metric.AddField("test", 13)
	fmt.Println(metric.String()) // added new value
	metric.RemoveField("test")
	fmt.Println(metric.String()) // remove old value, but colon error
	metric.Fields()              // panic because of colon from above
}
```

This creates the following output:
```
$ go test
m test="value" 1519041597145903000

m test="value" 1519041597145903000

m test="value",test=13i 1519041597145903000

m ,test=13i 1519041597145903000

--- FAIL: TestDirectModification (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 5 [running]:
testing.tRunner.func1(0xc4201080f0)
	/usr/local/Cellar/go/1.9.3/libexec/src/testing/testing.go:711 +0x2d2
panic(0x1292140, 0x1458a40)
	/usr/local/Cellar/go/1.9.3/libexec/src/runtime/panic.go:491 +0x283
github.com/influxdata/telegraf/metric.(*metric).Fields(0xc42009a200, 0x1)
	/Workspaces/go/src/github.com/influxdata/telegraf/metric/metric.go:345 +0xa22
github.com/influxdata/telegraf/plugins/processors/enum.TestDirectModification(0xc4201080f0)
	/Workspaces/go/src/github.com/influxdata/telegraf/plugins/processors/enum/demo_test.go:24 +0x4f9
testing.tRunner(0xc4201080f0, 0x12f5940)
	/usr/local/Cellar/go/1.9.3/libexec/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/Cellar/go/1.9.3/libexec/src/testing/testing.go:789 +0x2de
exit status 2
FAIL	github.com/influxdata/telegraf/plugins/processors/enum	0.020s
```

So the following behaviour can be observed:

- `RemoveField(test)` on the last field does not change the metric, the field remains.
- `AddField(test,13)` creates a new field with a different value
- `RemoveField(test)` removes the old value but leads to a leading ',' which indicates a problem
- the problem from above manifests in a panic when the `Fields()` is called 

This problem is always present, if the leading field in the string representation is removed. Even when more fields are present. Therefore, I decided, that modifying a field value directly by this approach is not possible. The issue lies in the unreliability of `RemoveField(...)`. 

Is there another approach to a direct modification? I really want to keep the feature of replacing the original value not just adding the mapped value as a new field.

Reply:
I'm rewriting the implementation of `telegraf.Metric` for 1.6 and I'll fix this, I'll let you know when its ready for testing.